### PR TITLE
report(psi): stub out locale swapping

### PR DESCRIPTION
**Pull request by paulirish**
Mostly meant as a recipe for our PSI implementation.

However, we add in these APIs to the report bundle. (and save >100kb by switching the lodash import)
and needed to adopt some of the rollup/fs magic from our other builds to avoid pulling in all the JSON into the build, etc.

### CODE REVIEW

File: shared/localization/swap-locale.js [hunk 1]
```diff
@@ -5,7 +5,9 @@
  */
 'use strict';
 
-const {set: _set, get: _get} = require('lodash');
+// Using these rather than the full 'lodash' saves 134kb to the report/bundle.umd.js
+const _get = require('lodash.get');
+const _set = require('lodash.set');
```
Comment (connorjclark) [inline shared/localization/swap-locale.js:10]: Did it really?

When I reviewed #13695 I checked for this and confirmed that all the extra lodash stuff was getting removed from the bundle.

File: report/test-assets/faux-psi.js [hunk 2]
```diff
@@ -82,6 +87,22 @@ async function renderLHReport() {
   }
 }
 
+/**
+ * @param {LH.Locale} locale
+ */
+async function swapLhrLocale(locale) {
+  const response = await fetch(`https://www.gstatic.com/pagespeed/insights/ui/locales/${locale}.json`);
```
Comment (brendankenny) [inline report/test-assets/faux-psi.js:94]: In the real thing, is the list of locales suggested in the UI generated from the actual locales available so it's not going to 404?

File: report/clients/bundle.js [hunk 3]
```diff
@@ -15,3 +15,4 @@ export {DOM} from '../renderer/dom.js';
 export {ReportRenderer} from '../renderer/report-renderer.js';
 export {ReportUIFeatures} from '../renderer/report-ui-features.js';
 export {renderReport} from '../renderer/api.js';
+export {swapLocale, format} from '../../shared/localization/i18n-module';
```
Comment (brendankenny) [inline report/clients/bundle.js:18]: what uses the esm report? I forget, but it uses `bundle.js` file too so it will have these in there now

Comment (paulirish) [reply report/clients/bundle.js:18]: DevTools uses bundle.esm.js.  
 (old psi used a psi.js build that was also esm, but that's dead now. new psi uses bundle.umd.js)

File: report/test-assets/faux-psi.js (hunk 2 above)
Comment (paulirish) [reply report/test-assets/faux-psi.js:94]: ideally, but there's no certainty the locales obj we have matches what's published here. but.. it _should_.

I think what you may be pointing out is that we should do a `lookupLocale` if the locale code coming from PSI hasnt already underwent that.  At this point, I'm not sure, but i'll find out. Easier to find out once we can try out this PR. :)

File: shared/localization/swap-locale.js (hunk 1 above)
Comment (paulirish) [reply shared/localization/swap-locale.js:10]: > Did it really?

yup.

image

I also tried moving the two localization/*.js files to esm so the lodash import was esm for treeshaking and i didnt get any savings in that case either.

I did not know about this CVE thing though.

Comment (paulirish) [reply shared/localization/swap-locale.js:10]: > I did not know about this CVE thing though.

also annoying.. the [alternatives](https://github.com/lukeed/dset#related) to lodash's get/set i'm finding don't use the square bracket notation we've been using... eg `'audits[is-on-https].title'`. (They just use `'audits.is-on-https.title'`). makes migration kinda a pain

Comment (connorjclark) [reply shared/localization/swap-locale.js:10]: I just checked again and confirmed that PR _did not_ change the size, so something else is going on here.

image

File: report/clients/bundle.js (hunk 3 above)
Comment (connorjclark) [inline report/clients/bundle.js:18]:
```suggestion
export {swapLocale, format} from '../../shared/localization/i18n-module.js';
```

File: shared/localization/swap-locale.js (hunk 1 above)
Comment (connorjclark) [reply shared/localization/swap-locale.js:10]: But it did change the viewer i18n module: 
image

in hindsight none of this lodash changes touched a report, except the i18n stuff, but only for the module/optional report code.

Comment (connorjclark) [reply shared/localization/swap-locale.js:10]: yeah the dt bundle increased in size, which def. uses lodash. oops :( 

image

Comment (connorjclark) [reply shared/localization/swap-locale.js:10]: looking at the imported module for lodash...it's a big mess of IIFE so no way treeshaking would work

but i think we can use modules directly: `lodash/isEqual.js`. I'll try that and maybe open a pr.

Comment (connorjclark) [reply shared/localization/swap-locale.js:10]: #13914

File: report/clients/bundle.js (hunk 3 above)
Comment (connorjclark) [reply report/clients/bundle.js]: so this adds code to the devtool report bundle, when it isn't needed there. should we avoid that?

or should we add swap locale to LH panel :)

**Review by brendankenny — APPROVED** [review]

**Review by connorjclark — CHANGES_REQUESTED** [review]
flagging this comment

https://github.com/GoogleChrome/lighthouse/pull/13885#discussion_r859063448

File: report/clients/bundle.js (hunk 3 above)
Comment (paulirish) [reply report/clients/bundle.js]: addressed in [68935ef](https://github.com/GoogleChrome/lighthouse/pull/13885/commits/68935ef1a8f03de861b3c5da277d09db11ea21f6)

File: report/test-assets/faux-psi.js (hunk 2 above)
Comment (paulirish) [reply report/test-assets/faux-psi.js:94]: > flagging this comment
> 
> [#13885 (comment)](https://github.com/GoogleChrome/lighthouse/pull/13885#discussion_r859063448)

addressed. @connorjclark  ^

**Comment by paulirish** [comment]
> flagging this comment
> 
> [#13885 (comment)](https://github.com/GoogleChrome/lighthouse/pull/13885#discussion_r859063448)

now addressed. @connorjclark  ^

File: build/build-report.js [hunk 4]
```diff
@@ -96,6 +96,11 @@ async function buildEsModulesBundle() {
     input: 'report/clients/bundle.js',
     plugins: [
       rollupPlugins.commonjs(),
+      // Exclude this 30kb from the devtools bundle for now.
+      rollupPlugins.shim({
+        [`${LH_ROOT}/shared/localization/i18n-module.js`]:
+            'export const swapLocale = _ => {}; export const format = _ => {};',
```
Comment (connorjclark) [inline build/build-report.js:102]: Can you do what build-bundle does instead? https://github.com/GoogleChrome/lighthouse/blob/master/build/build-bundle.js#L112

Comment (connorjclark) [reply build/build-report.js:102]: Oh, I guess you need some dummy functions exported here.

File: build/build-report.js [hunk 5]
```diff
@@ -110,19 +115,27 @@ async function buildUmdBundle() {
...
     input: 'report/clients/bundle.js',
     plugins: [
+      rollupPlugins.inlineFs({verbose: true}),
       rollupPlugins.commonjs(),
       rollupPlugins.terser({
         format: {
           beautify: true,
         },
       }),
+      // Shim this empty to ensure the bundle isn't 10MB
```
Comment (connorjclark) [inline build/build-report.js:125]: maybe we should have a test for this

**Review by connorjclark — APPROVED** [review]

File: build/build-report.js (hunk 5 above)
Comment (brendankenny) [reply build/build-report.js:125]: > maybe we should have a test for this

fry squinting

**Pull request merged**